WordPress/gutenberg
# Inner blocks: Reduce tree depth to improve performance.

## Related PRs:

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5746

## What?

Reduce the rendering complexity and depth of the block list for inner blocks.

## Why?

Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5617. 

Reducing rendering complexity is likely to improve performance overall. Also,
React Native has shown it has a limit in regards to number of nested views it
can render. Rendering too deep of trees can result in crashes; we observed this
when List v2 began using inner blocks.

## How?

- Combine `BlockList` and `BlockListCompact` to reduce code duplication between
  the separate components, e.g. item render, footer, end-of-list block appender
  button, empty list placeholder.
- Prepare support for additional block types to use the non-virtualized list
  through the addition of features like the nested block appender.

The following [diagram](https://excalidraw.com/#json=oXqmzQFp0lERueKKFVowL,RtdT1z4t0wFbkDG7yPbpXg) visually showcases the new structure of the block list used for 
both the root and inner blocks. 
![block-list-refactor](https://github.com/WordPress/gutenberg/assets/438664/03db4620-b02a-4fcb-9f2b-9aaa16d2ee28)


## Testing Instructions




We should test adding, editing, re-ordering, and removing various block types,
including nested blocks. Additionally, we should test various device
orientations and sizes.

Specifically, we should heavily test various organizations of inner blocks with 
different alignments applied to the parent and children blocks. The testing 
instructions for a couple of Columns block PRs are a good start: https://github.com/WordPress/gutenberg/pull/19013, 
https://github.com/WordPress/gutenberg/pull/25621. 

### Testing Instructions for Keyboard

n/a

## Screenshots or screencast 
n/a
